avm/src/vm/opcode.rs

```diff
@@ -3831,6 +3831,339 @@ pub static OPCODES: Lazy<HashMap<i64, ByteOpcode>> = Lazy::new(|| {
+            HandlerMemory::transfer(&hm, CLOSURE_ARG_MEM_START, &mut hand_mem, CLOSURE_ARG_MEM_START)?;
+            hand_mem.push_fixed(args[2], 1i64)?;
+            hand_mem.push_register(args[2], CLOSURE_ARG_MEM_START)?;
+          },
+          None => {
+            hand_mem.push_fixed(args[2], 0i64)?;
+            hand_mem.push_fractal(args[2], HandlerMemory::str_to_fractal("namespace-key pair not found"))?;
+          },
+        }
+      } else {
+        // TODO
+      }
+      Ok(hand_mem)
+    })
+  });
+  unpred_cpu!(dsmrun => fn(args, mut hand_mem) {
+    Box::pin(async move {
+      let nsref = hand_mem.read_fractal(args[0])?;
+      let ns = HandlerMemory::fractal_to_string(hand_mem.read_from_fractal(&nsref, 0).0)?;
+      let key = HandlerMemory::fractal_to_string(hand_mem.read_from_fractal(&nsref, 1).0)?;
+      let nskey = format!("{}:{}", ns, key);
+      let ctrl_port = CONTROL_PORT_CHANNEL.get();
+      let ctrl_port = match ctrl_port {
+        Some(ctrl_port) => Some(ctrl_port.borrow().clone()), // TODO: Use thread-local storage
+        None => None,
+      };
+      let is_key_owner = match ctrl_port {
+        Some(ref ctrl_port) => ctrl_port.is_key_owner(&nskey),
+        None => true,
+      };
+      if is_key_owner {
+        hand_mem.init_fractal(args[2])?;
+        let maybe_hm = DS.get(&nskey);
+        match maybe_hm {
+          Some(ds) => {
+            let mut hm = HandlerMemory::fork(hand_mem.clone())?;
+            HandlerMemory::transfer(&ds, 0, &mut hm, CLOSURE_ARG_MEM_START + 1)?;
+            let subhandler = HandlerFragment::new(args[1], 0);
+            let hm = subhandler.run(hm).await?;
+            let hm = hm.drop_parent()?;
+            HandlerMemory::transfer(&hm, CLOSURE_ARG_MEM_START, &mut hand_mem, CLOSURE_ARG_MEM_START)?;
+            hand_mem.push_fixed(args[2], 1i64)?;
+            hand_mem.push_register(args[2], CLOSURE_ARG_MEM_START)?;
+            // Also grab the mutation to the datastore value and re-insert it
+            let mut newds = HandlerMemory::new(None, 1)?;
+            HandlerMemory::transfer(&hm, CLOSURE_ARG_MEM_START + 1, &mut newds, 0)?;
+            drop(ds);
+            DS.insert(nskey, newds);
+          },
+          None => {
+            hand_mem.push_fixed(args[2], 0i64)?;
+            hand_mem.push_fractal(args[2], HandlerMemory::str_to_fractal("namespace-key pair not found"))?;
+          },
+        }
+      } else {
+        // TODO
+      }
+      Ok(hand_mem)
+    })
+  });
+  unpred_cpu!(dsrwith => fn(args, mut hand_mem) {
+    Box::pin(async move {
+      let with = hand_mem.read_fractal(args[0])?;
+      let nsref = hand_mem.read_from_fractal(&with, 0).0;
+      let ns = HandlerMemory::fractal_to_string(hand_mem.read_from_fractal(&nsref, 0).0)?;
+      let key = HandlerMemory::fractal_to_string(hand_mem.read_from_fractal(&nsref, 1).0)?;
+      let nskey = format!("{}:{}", ns, key);
+      hand_mem.register_out(args[0], 1, CLOSURE_ARG_MEM_START + 2)?;
```

Comment:
What does this `register_out` is doing?

Reply:
So this is a "with" opcode, so the value that's being passed in as the second parameter is nested as the second element of the `With<NsRef, any>` generic. Since all of our opcodes operate on the top-level addresses and have no nested pathing, this memory call allows us to take the value in an array offset and make a pointer to it in a top-level address.

It is literally identical to the [`register` opcode](https://github.com/alantech/alan/blob/main/avm/src/vm/opcode.rs#L2592-L2599) (which still has that old naming scheme because we didn't want to break compilation / change the compiler, but maybe we should just do that). There are three of these `HandlerMemory` methods:

1. `register` just creates a pointer from address A to address B
2. `register_out` creates a pointer in an array in address A at offset X to address B
3. `register_in` creates a pointer from address A to an array in address B at offset X

I hope that helps!